File: Adm/Models/src/PeriodoClima.java
package Adm.Models.src;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.HashMap;
import java.util.Map;

public class PeriodoClima implements Inter{
  private int id;
  private String cidade;
  private String pais;
  private String clima;
  private LocalDateTime data;
  private String periodo;
  private double temperatura;
  private double sensacaoTermica;

  public PeriodoClima(int id, String cidade) {
    setId(id);
    setCidade(cidade);
  }

  @Override
  public String toString() {
    return "Periodo_clima:" +
             "id=" + this.id +
             ", cidade=" + this.cidade;
  }

  public Map<String, Object> toDict() {
    Map<String, Object> dict = new HashMap<>();
    dict.put("id", this.id);
    dict.put("cidade", this.cidade);
    dict.put("pais", this.pais);
    dict.put("data", this.data.format(DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm")));
    dict.put("clima",this.clima);
    dict.put("periodo",this.periodo);
    dict.put("temperatura",this.temperatura);
    dict.put("sensacao_termica",this.sensacaoTermica);
    return dict;
}

  public void setId(int id){
    if (id > 0 ){
      this.id = id;
    }
    else{
      throw new IllegalArgumentException("id clima inválido");
    }
  }

  public void setCidade(String cidade) {
    if (cidade == null || cidade.isEmpty()) {
      throw new IllegalArgumentException("Cidade inválida");
    }
    this.cidade = cidade;
  }

  public int getId() {
    return this.id;
  }

  public String getCidade() {
    return this.cidade;
  }

  public String getClima() {
    return this.clima;
  }

  public LocalDateTime getData() {
    return this.data;
  }

  public String getPeriodo() {
    return this.periodo;
  }

  public void climaCidade() {
    String url = String.format(
        "http://api.openweathermap.org/data/2.5/weather?q=%s&appid=c5c2e778103223f1d989ac03ad6fcaee&units=metric&lang=pt_br",
        this.cidade);
    HttpRequest request = HttpRequest.newBuilder()
        .uri(URI.create(url))
        .GET()
        .build();

    try {
      HttpResponse<String> response = HttpClient.newHttpClient().send(request, HttpResponse.BodyHandlers.ofString());
      if (response.statusCode() == 200) {
        String dados = response.body();
        // Parse JSON
        this.pais = getJsonValue(dados, "sys.country");
        this.clima = getJsonValue(dados, "weather[0].description");
        this.temperatura = Double.parseDouble(getJsonValue(dados, "main.temp"));
        this.sensacaoTermica = Double.parseDouble(getJsonValue(dados, "main.feels_like"));
        this.data = LocalDateTime.now();
        this.periodo = determinarPeriodo();
      } else {
        this.clima = "indisponível";
        this.temperatura = Double.NaN;
        this.sensacaoTermica = Double.NaN;
        this.data = null;
        this.periodo = "indisponível";
      }
    } catch (IOException | InterruptedException e) {
      System.err.println("Erro ao consultar API: " + e.getMessage());
    }
  }

  private String getJsonValue(String json, String key) {
    int startIndex = json.indexOf(key) + key.length() + 4;
    int endIndex = json.indexOf(',', startIndex);
    return json.substring(startIndex, endIndex).trim().replaceAll("\"", "");
  }

  private String determinarPeriodo() {
    int mes = data.getMonthValue();
    if (mes >= 12 || mes <= 2)
      return "Inverno";
    else if (mes >= 3 && mes <= 5)
      return "Primavera";
    else if (mes >= 6 && mes <= 8)
      return "Verão";
    else
      return "Outono";
  }
}

File: Adm/Models/Periodo_Clima.cs
// using System;
// using System.Collections.Generic;
// using System.IO;
// using Newtonsoft.Json;
// using System.Net.Http;
// using System.Threading.Tasks;

// namespace Adm.Models{
//     public class PeriodoClima : Inter
//     {
//         private int id;
//         private string cidade;

//         public PeriodoClima(int id, string cidade)
//         {
//             setId(id);
//             setCidade(cidade);
//         }

//         public override string ToString()
//         {
//             return $"Periodo_clima:\nid={this.id}\n, cidade={this.cidade}";
//         }

//         public Dictionary<string, object> ToDict()
//             {
//                 return new Dictionary<string, object>
//                 {
//                     { "id", this.id },
//                     { "cidade", this.cidade },
//                     { "pais", this.pais },
//                     { "data", this.data.ToString("dd/MM/yyyy HH:mm") },
//                     { "clima", this.clima },
//                     { "periodo", this.periodo },
//                     { "temperatura", this.temperatura },
//                     { "sensacao_termica", this.sensacaoTermica }
//                 };
//             }

//         public int getId() => this.id;
//         public void setId(int id)
//         {
//             if (id >= 0)
//             {
//                 this.id = id;
//             }
//             else
//             {
//                 throw new ArgumentException("ID clima inválido");
//             }
//         }

//         public void setCidade(string cidade)
//         {
//             if (!string.IsNullOrEmpty(cidade))
//             {
//                 this.cidade = cidade;
//             }
//             else
//             {
//                 throw new ArgumentException("Cidade inválida");
//             }

//         }
        


//         public async Task ClimaCidadeAsync()
//         {
//             string url = $"http://api.openweathermap.org/data/2.5/weather?q={this.cidade}&appid=c5c2e778103223f1d989ac03ad6fcaee&units=metric&lang=pt_br";

//             using (HttpClient client = new HttpClient())
//             {
//                 HttpResponseMessage resposta = await client.GetAsync(url);

//                 if (resposta.IsSuccessStatusCode)
//                 {
//                     var dados = JsonConvert.DeserializeObject<Dictionary<string, object>>(await resposta.Content.ReadAsStringAsync());
//                     var sys = JsonConvert.DeserializeObject<Dictionary<string, object>>(dados["sys"].ToString());
//                     var weather = JsonConvert.DeserializeObject<List<Dictionary<string, object>>>(dados["weather"].ToString());
//                     var main = JsonConvert.DeserializeObject<Dictionary<string, object>>(dados["main"].ToString());

//                     this.pais = sys["country"].ToString();
//                     this.clima = weather[0]["description"].ToString();
//                     this.temperatura = Convert.ToSingle(main["temp"]);
//                     this.sensacaoTermica = Convert.ToSingle(main["feels_like"]);
//                     this.data = DateTime.Now;
//                     this.periodo = DeterminarPeriodo();
//                 }
//                 else
//                 {
//                     this.clima = "indisponível";
//                     this.temperatura = -1;
//                     this.sensacaoTermica = -1;
//                     this.data = DateTime.MinValue;
//                     this.periodo = "indisponível";
//                 }
//             }
//         }

//         private string DeterminarPeriodo()
//         {
//             int mes = data.Month;

//             if (mes == 12 || mes == 1 || mes == 2)
//                 return "Inverno";
//             if (mes >= 3 && mes <= 5)
//                 return "Primavera";
//             if (mes >= 6 && mes <= 8)
//                 return "Verão";
//             return "Outono";
//         }
//         public string getCidade() => this.cidade;
//         public string getClima() => this.clima;
//         public DateTime getData() => this.data;
//         public string getPais() => this.pais;
//         public float getTemperatura() => this.temperatura;
//         public string getPeriodo() => this.periodo;
//         public float getSensacaoTermica() => this.sensacaoTermica;
//     }

// }
